### ISSUE TITLE
Question : Adversarial Speaker Classifier

### ISSUE THREAD

**Issue by sooftware** [Tomiinek/Multilingual_Text_to_Speech#24]
Thank you for your awesome paper & code.  
I have a question about adversarial speaker classifier.  
   
I think, In training, the speaker classifier doesn't affect to rest model.  
In training, The speaker classifier and the rest of the model are independent of each other. Right?  
  
If my understanding right, What is the speaker classifier used for?
  
Thank you.

**Comment by Tomiinek** [comment]
Hello, thank you.

Well, I do not think so. The speaker classifier and the rest share the text encoder.
During the forward pass, the speaker classifier and the decoder are independent, but when backpropagating gradients, both losses (L2 of the predicted spectrogram and cross entropy of predicted speaker IDs) affect parameters of the encoder.

It is used, because it (thanks to the gradient reversal layer) suppresses speaker information leaking from the decoder to the text encoder (ideally, you should not be able to predict speaker from language). It seems that this helps when cloning voices across languages.

**Comment by sooftware** [comment]
Thank you for your answer !!   

I confuse role of adversarial speaker classifier and speaker embedding.  

Adversarial speaker classifier remove speaker-specific information from encoder outputs. But, isn't speaker embedding giving speaker-specific information??  

These two opposing roles are causing confusion in my head.   
Can you explain more detail??

Ah, Never mind the mail I sent you.

**Comment by sooftware** [comment]
> Hello, thank you.
> 
> Well, I do not think so. The speaker classifier and the rest share the text encoder.
> During the forward pass, the speaker classifier and the decoder are independent, but when backpropagating gradients, both losses (L2 of the predicted spectrogram and cross entropy of predicted speaker IDs) affect parameters of the encoder.
> 
> It is used, because it (thanks to the gradient reversal layer) suppresses speaker information leaking from the decoder to the text encoder (ideally, you should not be able to predict speaker from language or phoneme). It seems that this helps when cloning voices across languages.

Wow.. Your answer gave me a new insight. Thank you !!

**Comment by sooftware** [comment]
Oh, you're saying that the speaker information can't be included in the text encoder, right?  
Speaker information is provided by embedding from the speaker ID! Is that the difference?

**Comment by Tomiinek** [comment]
Yes, exactly! :smile: 

The encoded text is classified at first, then a speaker embedding is concatenated to it (in order to add the speaker information) and then passed into the decoder.

**Comment by sooftware** [comment]
Thank you for your answer.  
Thank you so much.  
Thanks to you, I was able to understand accurately.
